File: src/compiler/syntax/parser.rs
use crate::{
    compiler::lexer::{token_kind::TokenKind, token_stream::TokenStream},
    error::kaori_error::KaoriError,
    kaori_error,
};

use super::{
    ast_node::ASTNode,
    declaration::{Decl, DeclKind},
    expression::Expr,
    operator::{BinaryOp, UnaryOp},
    r#type::Type,
    statement::{Stmt, StmtKind},
};

pub struct Parser {
    token_stream: TokenStream,
}

impl Parser {
    pub fn new(token_stream: TokenStream) -> Self {
        Self { token_stream }
    }

    pub fn declarations(&mut self) -> Result<Vec<ASTNode>, KaoriError> {
        let mut declarations: Vec<ASTNode> = Vec::new();

        while !self.token_stream.at_end() {
            let declaration = self.parse_declaration()?;
            declarations.push(declaration);
        }

        Ok(declarations)
    }

    /* Declarations */
    fn parse_declaration(&mut self) -> Result<ASTNode, KaoriError> {
        let declaration = match self.token_stream.token_kind() {
            TokenKind::Function => self.parse_function_declaration()?,
            _ => {
                if !self
                    .token_stream
                    .look_ahead(&[TokenKind::Identifier, TokenKind::Colon])
                {
                    return Ok(ASTNode::Statement(self.parse_statement()?));
                }

                let variable_declaration = self.parse_variable_declaration()?;
                self.token_stream.consume(TokenKind::Semicolon)?;
                variable_declaration
            }
        };

        Ok(ASTNode::Declaration(declaration))
    }

    fn parse_variable_declaration(&mut self) -> Result<Decl, KaoriError> {
        let span = self.token_stream.span();
        let name = self.token_stream.lexeme();

        self.token_stream.consume(TokenKind::Identifier)?;
        self.token_stream.consume(TokenKind::Colon)?;

        let type_annotation = self.parse_type()?;

        self.token_stream.consume(TokenKind::Assign)?;

        let right = self.parse_expr()?;

        return Ok(Decl::variable(name, right, type_annotation, span));
    }

    fn parse_function_declaration(&mut self) -> Result<Decl, KaoriError> {
        let span = self.token_stream.span();

        self.token_stream.consume(TokenKind::Function)?;

        let name = self.token_stream.lexeme();

        self.token_stream.consume(TokenKind::Identifier);

        self.token_stream.consume(TokenKind::LeftParen);

        let mut parameters: Vec<Decl> = Vec::new();

        while !self.token_stream.at_end() && self.token_stream.token_kind() != TokenKind::RightParen
        {
            let parameter = self.parse_variable_declaration()?;
            parameters.push(parameter);

            if self.token_stream.token_kind() == TokenKind::RightParen {
                break;
            }

            self.token_stream.consume(TokenKind::Comma)?;
        }

        self.token_stream.consume(TokenKind::RightParen)?;
        self.token_stream.consume(TokenKind::ThinArrow)?;

        let return_type = self.parse_type()?;

        let block = self.parse_block_statement()?;

        let type_annotation = Type::Function {
            parameters: parameters
                .iter()
                .map(|param| match &param.kind {
                    DeclKind::Variable {
                        type_annotation, ..
                    } => type_annotation.clone(),
                    _ => unreachable!(),
                })
                .collect(),
            return_type: Box::new(return_type),
        };

        Ok(Decl::function(
            name,
            parameters,
            block,
            type_annotation,
            span,
        ))
    }

    /* Statements */
    fn parse_statement(&mut self) -> Result<Stmt, KaoriError> {
        let statement = match self.token_stream.token_kind() {
            TokenKind::Print => self.parse_print_statement(),
            TokenKind::LeftBrace => self.parse_block_statement(),
            TokenKind::If => self.parse_if_statement(),
            TokenKind::While => self.parse_while_loop_statement(),
            TokenKind::For => self.parse_for_loop_statement(),
            _ => self.parse_expr_statement(),
        }?;

        match statement.kind {
            StmtKind::Print(..) | StmtKind::Expression(..) => {
                self.token_stream.consume(TokenKind::Semicolon)?
            }
            _ => (),
        };

        return Ok(statement);
    }

    fn parse_expr_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();
        let expression = self.parse_expr()?;

        return Ok(Stmt::expression(expression, span));
    }

    fn parse_print_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();

        self.token_stream.consume(TokenKind::Print)?;
        self.token_stream.consume(TokenKind::LeftParen)?;
        let expression = self.parse_expr()?;
        self.token_stream.consume(TokenKind::RightParen)?;

        return Ok(Stmt::print(expression, span));
    }

    fn parse_block_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();

        let mut declarations: Vec<ASTNode> = Vec::new();

        self.token_stream.consume(TokenKind::LeftBrace)?;

        while !self.token_stream.at_end() && self.token_stream.token_kind() != TokenKind::RightBrace
        {
            let declaration = self.parse_declaration()?;
            declarations.push(declaration);
        }

        self.token_stream.consume(TokenKind::RightBrace)?;

        return Ok(Stmt::block(declarations, span));
    }

    fn parse_if_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();

        self.token_stream.consume(TokenKind::If)?;

        let condition = self.parse_expr()?;

        let then_branch = Box::new(self.parse_block_statement()?);

        if self.token_stream.token_kind() != TokenKind::Else {
            return Ok(Stmt::if_(condition, then_branch, None, span));
        }

        self.token_stream.advance();

        if self.token_stream.token_kind() == TokenKind::If {
            let else_branch = Some(Box::new(self.parse_if_statement()?));

            return Ok(Stmt::if_(condition, then_branch, else_branch, span));
        }

        let else_branch = Some(Box::new(self.parse_block_statement()?));

        return Ok(Stmt::if_(condition, then_branch, else_branch, span));
    }

    fn parse_while_loop_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();

        self.token_stream.consume(TokenKind::While)?;

        let condition = self.parse_expr()?;
        let block = Box::new(self.parse_block_statement()?);

        return Ok(Stmt::while_loop(condition, block, span));
    }

    fn parse_for_loop_statement(&mut self) -> Result<Stmt, KaoriError> {
        let span = self.token_stream.span();

        self.token_stream.consume(TokenKind::For)?;

        let declaration = self.parse_variable_declaration()?;

        self.token_stream.consume(TokenKind::Semicolon)?;

        let condition = self.parse_expr()?;

        self.token_stream.consume(TokenKind::Semicolon)?;

        let increment = self.parse_expr_statement()?;

        let mut block = self.parse_block_statement()?;

        if let StmtKind::Block(declarations) = &mut block.kind {
            declarations.push(ASTNode::Statement(increment));
        }

        let while_loop = Stmt::while_loop(condition, Box::new(block), span);

        let mut declarations: Vec<ASTNode> = Vec::new();

        declarations.push(ASTNode::Declaration(declaration));
        declarations.push(ASTNode::Statement(while_loop));

        return Ok(Stmt::block(declarations, span));
    }

    /* Exprs */
    fn parse_expr(&mut self) -> Result<Box<Expr>, KaoriError> {
        if self
            .token_stream
            .look_ahead(&[TokenKind::Identifier, TokenKind::Assign])
        {
            return self.parse_assign();
        }

        return self.parse_or();
    }

    fn parse_assign(&mut self) -> Result<Box<Expr>, KaoriError> {
        let identifier = self.parse_identifier()?;

        let span = self.token_stream.span();
        self.token_stream.consume(TokenKind::Assign)?;

        let right = self.parse_expr()?;

        return Ok(Box::new(Expr::assign(identifier, right, span)));
    }

    fn parse_or(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_and()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            if kind != TokenKind::Or {
                break;
            }

            self.token_stream.advance();

            let right = self.parse_and()?;

            left = Box::new(Expr::binary(BinaryOp::Or, left, right, span));
        }

        Ok(left)
    }

    fn parse_and(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_equality()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            if kind != TokenKind::And {
                break;
            }

            self.token_stream.advance();
            let right = self.parse_equality()?;

            left = Box::new(Expr::binary(BinaryOp::And, left, right, span));
        }

        Ok(left)
    }

    fn parse_equality(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_comparison()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            let operator = match kind {
                TokenKind::Equal => BinaryOp::Equal,
                TokenKind::NotEqual => BinaryOp::NotEqual,
                _ => break,
            };

            self.token_stream.advance();
            let right = self.parse_comparison()?;

            left = Box::new(Expr::binary(operator, left, right, span));
        }

        Ok(left)
    }

    fn parse_comparison(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_term()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            let operator = match kind {
                TokenKind::Greater => BinaryOp::Greater,
                TokenKind::GreaterEqual => BinaryOp::GreaterEqual,
                TokenKind::Less => BinaryOp::Less,
                TokenKind::LessEqual => BinaryOp::LessEqual,
                _ => break,
            };

            self.token_stream.advance();
            let right = self.parse_term()?;

            left = Box::new(Expr::binary(operator, left, right, span));
        }

        Ok(left)
    }

    fn parse_term(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_factor()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            let operator = match kind {
                TokenKind::Plus => BinaryOp::Plus,
                TokenKind::Minus => BinaryOp::Minus,
                _ => break,
            };

            self.token_stream.advance();
            let right = self.parse_factor()?;

            left = Box::new(Expr::binary(operator, left, right, span));
        }

        Ok(left)
    }

    fn parse_factor(&mut self) -> Result<Box<Expr>, KaoriError> {
        let mut left = self.parse_unary()?;

        while !self.token_stream.at_end() {
            let kind = self.token_stream.token_kind();
            let span = self.token_stream.span();

            let operator = match kind {
                TokenKind::Multiply => BinaryOp::Multiply,
                TokenKind::Divide => BinaryOp::Divide,
                TokenKind::Remainder => BinaryOp::Remainder,
                _ => break,
            };

            self.token_stream.advance();
            let right = self.parse_unary()?;

            left = Box::new(Expr::binary(operator, left, right, span))
        }

        return Ok(left);
    }

    fn parse_unary(&mut self) -> Result<Box<Expr>, KaoriError> {
        let kind = self.token_stream.token_kind();
        let span = self.token_stream.span();

        let operator = match kind {
            TokenKind::Plus => {
                self.token_stream.advance();
                return self.parse_unary();
            }
            TokenKind::Minus => UnaryOp::Negate,
            TokenKind::Not => UnaryOp::Not,
            _ => return self.parse_primary(),
        };

        self.token_stream.advance();

        Ok(Box::new(Expr::unary(operator, self.parse_unary()?, span)))
    }

    fn parse_primary(&mut self) -> Result<Box<Expr>, KaoriError> {
        let kind = self.token_stream.token_kind();
        let span = self.token_stream.span();

        Ok(match kind {
            TokenKind::LeftParen => {
                self.token_stream.consume(TokenKind::LeftParen)?;
                let expr = self.parse_expr()?;
                self.token_stream.consume(TokenKind::RightParen)?;
                expr
            }
            TokenKind::NumberLiteral => {
                let lexeme = self.token_stream.lexeme();
                let value = lexeme.parse::<f64>().unwrap();

                self.token_stream.advance();
                Box::new(Expr::number_literal(value, span))
            }
            TokenKind::BooleanLiteral => {
                let lexeme = self.token_stream.lexeme();
                let value = lexeme.parse::<bool>().unwrap();

                self.token_stream.advance();
                Box::new(Expr::boolean_literal(value, span))
            }
            TokenKind::StringLiteral => {
                let lexeme = self.token_stream.lexeme();

                self.token_stream.advance();
                Box::new(Expr::string_literal(lexeme, span))
            }
            TokenKind::Identifier => self.parse_identifier()?,
            _ => {
                let span = self.token_stream.span();
                return Err(kaori_error!(span, "{:?} is a invalid operand", kind));
            }
        })
    }

    fn parse_identifier(&mut self) -> Result<Box<Expr>, KaoriError> {
        let name = self.token_stream.lexeme();
        let span = self.token_stream.span();

        let identifier = Box::new(Expr::identifier(name, span));

        self.token_stream.consume(TokenKind::Identifier)?;

        Ok(identifier)
    }

    /* Types */
    pub fn parse_type(&mut self) -> Result<Type, KaoriError> {
        match self.token_stream.token_kind() {
            TokenKind::Identifier => self.parse_primitive_type(),
            _ => Err(kaori_error!(
                self.token_stream.span(),
                "invalid type annotation: {:?}",
                self.token_stream.token_kind(),
            )),
        }
    }

    fn parse_primitive_type(&mut self) -> Result<Type, KaoriError> {
        let sub = self.token_stream.lexeme();

        let primitive = match sub.as_str() {
            "bool" => Type::Boolean,
            "str" => Type::String,
            "number" => Type::Number,
            _ => {
                return Err(kaori_error!(
                    self.token_stream.span(),
                    "invalid primitive type: {:?}",
                    self.token_stream.token_kind(),
                ));
            }
        };

        self.token_stream.advance();

        Ok(primitive)
    }
}

File: legacy-java/src/main/java/com/kaori/compiler/syntax/Parser.java
package com.kaori.compiler.syntax;

import java.util.ArrayList;
import java.util.List;

import com.kaori.compiler.lexer.TokenKind;
import com.kaori.compiler.lexer.TokenStream;
import com.kaori.error.KaoriError;

public class Parser {
    private final TokenStream tokens;

    public Parser(TokenStream tokens) {
        this.tokens = tokens;
    }

    public List<DeclarationAST> declarations() {
        List<DeclarationAST> declarations = new ArrayList<>();

        while (!this.tokens.atEnd()) {
            DeclarationAST declaration = this.declaration();
            declarations.add(declaration);

        }

        return declarations;
    }

    /*
     * ─────────────────────────────
     * Declarations
     * ─────────────────────────────
     */
    private DeclarationAST declaration() {
        DeclarationAST declaration = switch (this.tokens.getCurrent()) {
            case FUNCTION -> this.functionDeclaration();
            default -> this.tokens.lookAhead(TokenKind.IDENTIFIER, TokenKind.COLON) ? this.variableDeclaration()
                    : this.statement();
        };

        if (declaration instanceof StatementAST.Print || declaration instanceof DeclarationAST.Variable
                || declaration instanceof StatementAST.Expr)

        {
            this.tokens.consume(TokenKind.SEMICOLON);
        }

        return declaration;
    }

    private DeclarationAST.Variable variableDeclaration() {
        int line = this.tokens.getLine();

        ExpressionAST.Identifier left = this.identifier();
        this.tokens.consume(TokenKind.COLON);

        TypeAST type = this.type();

        if (this.tokens.getCurrent() != TokenKind.ASSIGN) {
            ExpressionAST right = new ExpressionAST.Literal(type, null);
            return new DeclarationAST.Variable(line, left, right, type);
        }

        this.tokens.consume(TokenKind.ASSIGN);

        ExpressionAST right = this.expression();

        return new DeclarationAST.Variable(line, left, right, type);
    }

    private DeclarationAST functionDeclaration() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.FUNCTION);

        ExpressionAST.Identifier name = this.identifier();

        this.tokens.consume(TokenKind.LEFT_PAREN);

        List<DeclarationAST.Variable> parameters = new ArrayList<>();

        while (!this.tokens.atEnd() && this.tokens.getCurrent() != TokenKind.RIGHT_PAREN) {
            DeclarationAST.Variable parameter = this.variableDeclaration();
            parameters.add(parameter);

            if (this.tokens.getCurrent() == TokenKind.RIGHT_PAREN) {
                break;
            }

            this.tokens.consume(TokenKind.COMMA);
        }

        this.tokens.consume(TokenKind.RIGHT_PAREN);

        this.tokens.consume(TokenKind.THIN_ARROW);

        TypeAST returnType = this.type();

        TypeAST.Function type = new TypeAST.Function(parameters.stream().map(parameter -> parameter.type()).toList(),
                returnType);

        StatementAST.Block block = this.blockStatement();

        return new DeclarationAST.Function(line, name, parameters, type, block);
    }

    /*
     * ─────────────────────────────
     * Statements
     * ─────────────────────────────
     */
    private StatementAST.Expr expressionStatement() {
        int line = this.tokens.getLine();

        ExpressionAST expression = this.expression();

        return new StatementAST.Expr(line, expression);
    }

    private StatementAST.Print printStatement() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.PRINT);

        this.tokens.consume(TokenKind.LEFT_PAREN);

        ExpressionAST expression = this.expression();

        this.tokens.consume(TokenKind.RIGHT_PAREN);

        return new StatementAST.Print(line, expression);
    }

    private StatementAST.Block blockStatement() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.LEFT_BRACE);

        List<DeclarationAST> declarations = new ArrayList<>();

        while (!this.tokens.atEnd() && this.tokens.getCurrent() != TokenKind.RIGHT_BRACE) {
            DeclarationAST declaration = this.declaration();
            declarations.add(declaration);
        }

        this.tokens.consume(TokenKind.RIGHT_BRACE);

        return new StatementAST.Block(line, declarations);
    }

    private StatementAST.If ifStatement() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.IF);

        ExpressionAST condition = this.expression();

        StatementAST.Block thenBranch = this.blockStatement();

        if (this.tokens.getCurrent() != TokenKind.ELSE) {

            StatementAST.Block elseBranch = new StatementAST.Block(line);
            return new StatementAST.If(line, condition, thenBranch, elseBranch);
        }

        this.tokens.consume(TokenKind.ELSE);

        StatementAST elseBranch = switch (this.tokens.getCurrent()) {
            case IF -> this.ifStatement();
            default -> this.blockStatement();
        };

        return new StatementAST.If(line, condition, thenBranch, elseBranch);
    }

    private StatementAST.WhileLoop whileLoopStatement() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.WHILE);

        ExpressionAST condition = this.expression();

        StatementAST.Block block = this.blockStatement();

        return new StatementAST.WhileLoop(line, condition, block);
    }

    private StatementAST.Block forLoopStatement() {
        int line = this.tokens.getLine();

        this.tokens.consume(TokenKind.FOR);

        DeclarationAST.Variable variable = this.variableDeclaration();

        this.tokens.consume(TokenKind.SEMICOLON);

        ExpressionAST condition = this.expression();

        this.tokens.consume(TokenKind.SEMICOLON);

        StatementAST.Expr increment = this.expressionStatement();

        StatementAST.Block block = this.blockStatement();

        StatementAST.WhileLoop whileLoop = new StatementAST.WhileLoop(line, condition, block);

        List<DeclarationAST> declarations = new ArrayList<>();

        declarations.add(variable);
        declarations.add(whileLoop);

        whileLoop.block().declarations().add(increment);

        return new StatementAST.Block(line, declarations);
    }

    private StatementAST statement() {
        return switch (this.tokens.getCurrent()) {
            case PRINT -> this.printStatement();
            case IF -> this.ifStatement();
            case WHILE -> this.whileLoopStatement();
            case FOR -> this.forLoopStatement();
            default -> this.expressionStatement();
        };
    }

    /*
     * ─────────────────────────────
     * Expressions
     * ─────────────────────────────
     */
    private ExpressionAST expression() {
        if (this.tokens.lookAhead(TokenKind.IDENTIFIER, TokenKind.ASSIGN)) {
            return this.assign();
        }

        return this.or();
    }

    private ExpressionAST assign() {
        ExpressionAST.Identifier left = this.identifier();
        this.tokens.consume(TokenKind.ASSIGN);
        ExpressionAST right = this.expression();

        return new ExpressionAST.Assign(left, right);
    }

    private ExpressionAST or() {
        ExpressionAST left = this.and();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case OR -> {
                    this.tokens.consume();
                    ExpressionAST right = this.and();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.OR);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST and() {
        ExpressionAST left = this.equality();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case AND -> {
                    this.tokens.consume();
                    ExpressionAST right = this.equality();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.AND);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST equality() {
        ExpressionAST left = this.comparison();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case EQUAL -> {
                    this.tokens.consume();
                    ExpressionAST right = this.comparison();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.EQUAL);
                }
                case NOT_EQUAL -> {
                    this.tokens.consume();
                    ExpressionAST right = this.comparison();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.NOT_EQUAL);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST comparison() {
        ExpressionAST left = this.term();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case GREATER -> {
                    this.tokens.consume();
                    ExpressionAST right = this.term();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.GREATER);
                }
                case GREATER_EQUAL -> {
                    this.tokens.consume();
                    ExpressionAST right = this.term();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.GREATER_EQUAL);
                }
                case LESS -> {
                    this.tokens.consume();
                    ExpressionAST right = this.term();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.LESS);
                }
                case LESS_EQUAL -> {
                    this.tokens.consume();
                    ExpressionAST right = this.term();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.LESS_EQUAL);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST term() {
        ExpressionAST left = this.factor();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case PLUS -> {
                    this.tokens.consume();
                    ExpressionAST right = this.factor();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.PLUS);
                }
                case MINUS -> {
                    this.tokens.consume();
                    ExpressionAST right = this.factor();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.MINUS);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST factor() {
        ExpressionAST left = this.prefixUnary();

        while (!this.tokens.atEnd()) {
            TokenKind operator = this.tokens.getCurrent();

            switch (operator) {
                case MULTIPLY -> {
                    this.tokens.consume();
                    ExpressionAST right = this.prefixUnary();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.MULTIPLY);
                }
                case DIVIDE -> {
                    this.tokens.consume();
                    ExpressionAST right = this.prefixUnary();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.DIVIDE);
                }
                case MODULO -> {
                    this.tokens.consume();
                    ExpressionAST right = this.prefixUnary();
                    left = new ExpressionAST.BinaryExpression(left, right, ExpressionAST.BinaryOperator.MODULO);
                }
                default -> {
                    return left;
                }
            }
        }

        return left;
    }

    private ExpressionAST prefixUnary() {
        return switch (this.tokens.getCurrent()) {
            case MINUS -> {
                this.tokens.consume();
                yield new ExpressionAST.UnaryExpression(this.prefixUnary(), ExpressionAST.UnaryOperator.NEGATE);
            }
            case NOT -> {
                this.tokens.consume();
                yield new ExpressionAST.UnaryExpression(this.prefixUnary(), ExpressionAST.UnaryOperator.NOT);
            }
            case PLUS -> {
                this.tokens.consume();
                yield this.prefixUnary();

            }
            default -> this.primary();
        };
    }

    private ExpressionAST primary() {
        return switch (this.tokens.getCurrent()) {
            case BOOLEAN_LITERAL -> {
                String lexeme = this.tokens.getLexeme();
                boolean value = Boolean.parseBoolean(lexeme);
                ExpressionAST literal = new ExpressionAST.Literal(TypeAST.Primitive.BOOLEAN, value);
                this.tokens.consume();

                yield literal;
            }
            case STRING_LITERAL -> {
                String lexeme = this.tokens.getLexeme();
                String value = lexeme.substring(1, lexeme.length() - 1);
                TypeAST type = TypeAST.Primitive.STRING;
                ExpressionAST literal = new ExpressionAST.Literal(type, value);
                this.tokens.consume();

                yield literal;
            }
            case NUMBER_LITERAL -> {
                String lexeme = this.tokens.getLexeme();
                Double value = Double.parseDouble(lexeme);
                TypeAST type = TypeAST.Primitive.NUMBER;
                ExpressionAST literal = new ExpressionAST.Literal(type, value);
                this.tokens.consume();

                yield literal;
            }
            case IDENTIFIER -> {
                ExpressionAST.Identifier identifier = this.identifier();
                yield this.postfixUnary(identifier);

            }
            case LEFT_PAREN -> this.parenthesis();
            default -> throw KaoriError.SyntaxError("expected a valid operand", this.tokens.getLine());
        };
    }

    private ExpressionAST.Identifier identifier() {
        String lexeme = this.tokens.getLexeme();
        this.tokens.consume(TokenKind.IDENTIFIER);

        return new ExpressionAST.Identifier(lexeme);
    }

    private ExpressionAST parenthesis() {
        this.tokens.consume(TokenKind.LEFT_PAREN);

        ExpressionAST expression = this.expression();

        this.tokens.consume(TokenKind.RIGHT_PAREN);

        return expression;
    }

    private ExpressionAST postfixUnary(ExpressionAST.Identifier identifier) {
        return switch (this.tokens.getCurrent()) {
            case INCREMENT -> {
                ExpressionAST literal = new ExpressionAST.Literal(TypeAST.Primitive.NUMBER, 1.0);
                ExpressionAST add = new ExpressionAST.BinaryExpression(identifier, literal,
                        ExpressionAST.BinaryOperator.PLUS);
                this.tokens.consume();
                yield new ExpressionAST.Assign(identifier, add);
            }
            case DECREMENT -> {
                ExpressionAST literal = new ExpressionAST.Literal(TypeAST.Primitive.NUMBER, 1.0);
                ExpressionAST subtract = new ExpressionAST.BinaryExpression(identifier, literal,
                        ExpressionAST.BinaryOperator.MINUS);
                this.tokens.consume();
                yield new ExpressionAST.Assign(identifier, subtract);
            }
            case LEFT_PAREN -> this.functionCall(identifier);
            default -> identifier;
        };
    }

    private ExpressionAST functionCall(ExpressionAST callee) {
        if (this.tokens.getCurrent() != TokenKind.LEFT_PAREN) {
            return callee;
        }

        this.tokens.consume(TokenKind.LEFT_PAREN);

        List<ExpressionAST> arguments = new ArrayList<>();

        while (!this.tokens.atEnd() && this.tokens.getCurrent() != TokenKind.RIGHT_PAREN) {
            ExpressionAST argument = this.expression();
            arguments.add(argument);

            if (this.tokens.getCurrent() == TokenKind.RIGHT_PAREN) {
                break;
            }

            this.tokens.consume(TokenKind.COMMA);
        }

        this.tokens.consume(TokenKind.RIGHT_PAREN);

        ExpressionAST call = new ExpressionAST.FunctionCall(callee, arguments);

        return this.functionCall(call);
    }

    /*
     * ─────────────────────────────
     * Types
     * ─────────────────────────────
     */
    private TypeAST type() {
        TypeAST type = switch (this.tokens.getCurrent()) {
            case IDENTIFIER -> this.primitiveType();
            default ->
                throw KaoriError.SyntaxError(this.tokens.getCurrent() + " is not a valid type", this.tokens.getLine());
        };

        return type;
    }

    private TypeAST primitiveType() {
        String lexeme = this.tokens.getLexeme();
        this.tokens.consume(TokenKind.IDENTIFIER);

        TypeAST type = switch (lexeme) {
            case "str" -> TypeAST.Primitive.STRING;
            case "bool" -> TypeAST.Primitive.BOOLEAN;
            case "number" -> TypeAST.Primitive.NUMBER;
            case "void" -> TypeAST.Primitive.VOID;
            default ->
                throw KaoriError.SyntaxError(this.tokens.getCurrent() + " is not a valid type", this.tokens.getLine());
        };

        return type;
    }

}
